src/assert_safe.h

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+#ifndef INCLUDE_assert_safe_h__
+#define INCLUDE_assert_safe_h__
+
+/*
+ * In a debug build, we'll assert(3) for aide in debugging.  In release builds,
+ * we will provide macros that will set an error message that indicate a failure
+ * and return.  Note that memory leaks can occur in a release-mode assertion
+ * failure -- it is impractical to provide safe clean up routines in these very
+ * extreme failures, but care should be taken to not leak very large objects.
+ */
+
+#if (defined(_DEBUG) || defined(GIT_ASSERT_HARD)) && GIT_ASSERT_HARD != 0
```

Comment:
Shouldn't this rather be `#if defined(_DEBUG) || (defined(GIT_ASSERT_HARD) && GIT_ASSERT_HARD)`?

Reply:
And I think we should also include <assert.h>, right?

Reply:
I don't think so - `GIT_ASSERT_HARD` is meant to override `_DEBUG`.  So `GIT_ASSERT_HARD=0` will go to our error set and return -1 functions even when `_DEBUG` is set.  `GIT_ASSERT_HARD=1` will go to the system assert, even when `_DEBUG` is unset.

Reply:
Ok. The only case that makes me wonder is if `GIT_ASSERT_HARD` isn't defined at all, as I'm not quite sure how above expression would be evaluated. So can this actually happen?